### PR TITLE
Use a lock for `get` and `set` of EmbeddingTable

### LINKED ISSUE #1808: Exception calling application: dictionary changed size during iteration
The exception occasionally happens when the worker reports gradients to ps.
image

### PR DESCRIPTION
**Pull request by workingloong**
Fix #1808 
The parameter dictionary may be updated by other threads during iteration to save it into a checkpoint file.

### CODE REVIEW

File: elasticdl/python/ps/parameters.py [hunk 1]
```diff
@@ -175,10 +175,13 @@ def to_model_pb(self):
...
         model_pb.version = self.version
-        for name, var in self.non_embedding_params.items():
+        non_embedding_params = self.non_embedding_params.copy()
+        for name, var in non_embedding_params.items():
             serialize_ndarray(var.numpy(), model_pb.dense_parameters[name])
+        del non_embedding_params
 
-        for name, embedding_table in self.embedding_params.items():
+        embedding_params = self.embedding_params.copy()
+        for name, embedding_table in embedding_params.items():
```
Comment (skydoorkai) [inline elasticdl/python/ps/parameters.py]: Maybe this embedding_param iteration is not the cause of the bug, as `self.embedding_params` has a fixed number (number of embedding tables) after initialization.

It is probably in `embedding_table.to_indexed_slices(),` a few lines later, as when iterating through the embedding vectos, new vectors may be added by other threads.

Comment (workingloong) [reply elasticdl/python/ps/parameters.py]: Yes. It is more suitable to copy `embedding_table.embedding_vectors` before iteration or use a lock ?

Comment (skydoorkai) [reply elasticdl/python/ps/parameters.py]: embedding_table.embedding_vectors may be too long for copy. If lock does not affect training efficiency when checkpoint is not saved, better to use a lock.

Comment (workingloong) [reply elasticdl/python/ps/parameters.py]: In PS with Python, the `optimizer_wrapper` has used a lock to update the embedding table. https://github.com/sql-machine-learning/elasticdl/blob/e695ba7b33f01383798b28825fab408dae4e1327/elasticdl/python/ps/optimizer_wrapper.py#L155-L157

So, it does not affect training efficiency to use a lock for `set` and `get` in `EmbeddingTable`. I will use a lock in the next commit.

File: elasticdl/python/ps/embedding_table.py [hunk 2]
```diff
@@ -44,34 +46,38 @@ def __init__(self, name, dim=None, initializer=None, is_slot=False):
...
-        for i in indices:
-            value = self.embedding_vectors.get(i, None)
-            if value is None:
-                value = self.initializer(shape=(self.dim,)).numpy()
-                self.embedding_vectors[i] = value
-            values.append(value)
+        with self._lock:
+            for i in indices:
+                value = self.embedding_vectors.get(i, None)
+                if value is None:
```
Comment (skydoorkai) [inline elasticdl/python/ps/embedding_table.py]: Only adding embedding vector and iteration need the lock. Reading or updating existing vectors does not.
So Move the lock here after `if value is None:` for adding. 

Also, `def set(self, indices, values)` below  does not need the lock, as it does not add any vectors.

Comment (QiJune) [reply elasticdl/python/ps/embedding_table.py]: We need a RWLock here. Please refer to the implementation in Go PS. https://github.com/sql-machine-learning/elasticdl/blob/develop/elasticdl/pkg/common/embedding_table.go#L14

The lock is used at `GetEmbeddingVector ` func and `ToIndexedSlices ` func.

Comment (workingloong) [reply elasticdl/python/ps/embedding_table.py]: If we don't use lock the `def set(self, indices, values)`, Reading and writing the `embedding_vector` are simultaneous which may cause inconsistency.

Comment (QiJune) [reply elasticdl/python/ps/embedding_table.py]: @workingloong  Yes. I am not sure if the reading/writeing concurrently will hurt the model accuracy. It seems that it's a tradeoff here, training speed and model accuracy. We could do some expriments.

Comment (skydoorkai) [reply elasticdl/python/ps/embedding_table.py]: In async-SGD, reading/writing inconsistency does not matter, as the async-SGD itself has a staleness issue.
In sync-SGD, there won't be any reading/writing inconsistency.

Comment (workingloong) [reply elasticdl/python/ps/embedding_table.py]: OK, we may need to do some experiments to view training speed and model accuracy and then to add the lock. Now, I only use the lock for iteration and set new vectors.

**Review by skydoorkai — APPROVED** [review]

**Pull request merged**